src/index.ts

```diff
@@ -1,69 +1,290 @@
+      : filepath;
+
+  const headless = coerceBooleanString(queryParameters.headless);
+
+  const updatedQueryParameters = {
+    ...queryParameters,
+    headless: headless,
+    filepath: selectedFilepath
+  };
+
+  let mainAreaWidget: MainAreaWidget<DeployAppWidget> | undefined;
+
+  if (headless === 'true') {
+    try {
+      if (!mainAreaWidget || mainAreaWidget.isDisposed) {
+        const content = new DeployAppWidget({
+          baseUrl: baseUrl,
+          queryParameters: updatedQueryParameters
+        });
+        mainAreaWidget = new MainAreaWidget({ content });
+      }
+
+      if (!widgetTracker.has(mainAreaWidget)) {
+        widgetTracker.add(mainAreaWidget);
+      }
+
+      // non-empty origin implies user has called command e.g. main-menu click.
+      // guards against (what appears to be) command execution from tracker
+      // on session restore.
+      if (origin && !mainAreaWidget.isAttached) {
+        IShell.add(mainAreaWidget, 'main');
+      }
+    } catch (error) {
+      console.warn(`Error opening in headless mode: ${error}`);
+    }
+  } else {
+    try {
+      // no restore logic required here as opening in a separate window
+      const completeUrl = `${baseUrl}?${buildQueryString(updatedQueryParameters)}`;
+      window.open(completeUrl, '_blank', 'noopener,noreferrer');
+    } catch (error) {
+      console.warn(`Error opening in window: ${error}`);
+    }
+  }
+};
+
 const jhubAppsPlugin: JupyterFrontEndPlugin<void> = {
-  id: 'jupyterlab-jhub-apps:commands',
+  id: PLUGIN_ID,
   description: 'Adds additional commands used by jhub-apps.',
   autoStart: true,
-  activate: (app: JupyterFrontEnd) => {
-    const openURL = (url: string) => {
-      try {
-        window.open(url, '_blank', 'noopener,noreferrer');
-      } catch (error) {
-        console.warn(`Error opening ${url}: ${error}`);
-      }
-    };
-
-    const calculateIcon = (args: IDeployAppArgs) => {
-      switch (args.origin) {
-        case 'main-menu':
-          return undefined;
-        case 'context-menu':
-        case undefined:
-        default:
-          return deployAppIcon;
-      }
-    };
-
-    app.commands.addCommand(CommandIDs.deployApp, {
-      execute: () => {
-        const currentWidget = app.shell.currentWidget;
-        const currentNotebookPath =
-          currentWidget && currentWidget instanceof DocumentWidget
-            ? currentWidget.context.path
-            : '';
-        let deployUrl;
-        if (currentNotebookPath !== '') {
-          deployUrl = `/services/japps/create-app?filepath=${encodeURIComponent(currentNotebookPath)}`;
-        } else {
-          deployUrl = '/services/japps/create-app';
-        }
-        openURL(deployUrl);
-      },
+  requires: [ISettingRegistry],
+  optional: [ILayoutRestorer],
+  activate: async (
+    app: JupyterFrontEnd,
+    settingsRegistry: ISettingRegistry,
+    restorer: ILayoutRestorer | null
+  ) => {
+    const { commands, shell } = app;
+
+    try {
```

Comment:
Lines 205 - 211 are the new code that try to deal with when a user updates their settings.